File: scripts/update_git_settings.py
import subprocess


def check_git_config_setting(section, key):
    # Get the list of all settings in a given section
    result = subprocess.run(
        ["git", "config", "--get-all", f"{section}.{key}"], stdout=subprocess.PIPE
    )
    # Decode the result and split into lines
    fetch_settings = result.stdout.decode("utf-8").strip().split("\n")
    # Check if the setting is already in the config
    return fetch_settings


def add_git_config_setting(section, key, value):
    # Add the setting using git config --add
    subprocess.run(["git", "config", "--add", f"{section}.{key}", value])


# The setting we want to check and possibly add for remote.origin.fetch
refspec_setting = "^refs/heads/gh-pages"

# Check if the setting is already applied for remote.origin.fetch
if refspec_setting not in check_git_config_setting("remote.origin", "fetch"):
    print("Setting not found. Excluding gh-pages from git fetch...")
    add_git_config_setting("remote.origin", "fetch", refspec_setting)
else:
    print("Setting already applied. Skipping excluding gh-pages from git fetch...")

# The setting we want to check and possibly add for submodule.fetchJobs
submodule_section = "submodule"
submodule_key = "fetchJobs"
submodule_value = "10"

# Check if the setting is already applied for submodule.fetchJobs
if submodule_value not in check_git_config_setting(submodule_section, submodule_key):
    print(
        f"Setting not found. Setting {submodule_section}.{submodule_key} to {submodule_value}..."
    )
    add_git_config_setting(submodule_section, submodule_key, submodule_value)
else:
    print(
        f"Setting already applied. Skipping setting {submodule_section}.{submodule_key} to {submodule_value}..."
    )

File: lib/memo/update_git_settings.ex
defmodule Memo.UpdateGitSettings do
  defp check_git_config_setting(section, key) do
    case System.cmd("git", ["config", "--get-all", "#{section}.#{key}"]) do
      {result, 0} ->
        String.split(String.trim(result), "\n")

      # Setting does not exist
      {"", 1} ->
        []

      {error_message, _} ->
        IO.puts("Error checking git config: #{error_message}")
        []
    end
  end

  defp add_git_config_setting(section, key, value) do
    System.cmd("git", ["config", "--add", "#{section}.#{key}", value])
  end

  defp handle_git_setting(section, key, value, message) do
    settings = check_git_config_setting(section, key)

    if value not in settings do
      IO.puts("#{message} Setting #{section}.#{key} to #{value}...")
      add_git_config_setting(section, key, value)
    else
      IO.puts("Setting already applied. Skipping #{message} for #{section}.#{key}")
    end
  end

  def run do
    handle_git_setting(
      "remote.origin",
      "fetch",
      "^refs/heads/gh-pages",
      "Excluding gh-pages from git fetch"
    )

    handle_git_setting("submodule", "fetchJobs", "10", "Setting submodule fetch jobs")
  end
end
